src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.cs

```diff
@@ -182,6 +185,67 @@ public void GarbageCollect(Func<string, IWorkloadResolver> getResolverForWorkloa
             }
         }
 
+        public string InstallWorkloadSet(string path)
+        {
+            return ModifyWorkloadSet(path, InstallAction.Install);
+        }
+
+        public void RollBackWorkloadSetInstallation()
+        {
+            if (_newWorkloadSetPath is not null)
+            {
+                ModifyWorkloadSet(_newWorkloadSetPath, InstallAction.Uninstall);
+                if (_previousWorkloadSetVersion is not null)
+                {
+                    File.WriteAllText(Path.Combine(_newWorkloadSetPath, "version.txt"), _previousWorkloadSetVersion);
+                    ModifyWorkloadSet(_newWorkloadSetPath, InstallAction.Repair);
+                }
+            }
+        }
+
+        private string ModifyWorkloadSet(string path, InstallAction installAction)
+        {
+            ReportPendingReboot();
+
+            // Rolling back a manifest update after a successful install is essentially a downgrade, which is blocked so we have to
+            // treat it as a special case and is different from the install failing and rolling that back, though depending where the install
+            // failed, it may have removed the old product already.
+            // Resolve the package ID for the manifest payload package
+            var featureBand = Path.GetFileName(Path.GetDirectoryName(path));
+            var version = File.ReadAllText(Path.Combine(path, "version.txt"));
+            string msiPackageId = GetManifestPackageId(new ManifestId("Microsoft.NET.Workloads"), new SdkFeatureBand(featureBand)).ToString();
+            string msiPackageVersion = version;
+
+            Log?.LogMessage($"Resolving Microsoft.NET.Workloads ({version}) to {msiPackageId} ({msiPackageVersion}).");
+
+            // Retrieve the payload from the MSI package cache.
+            MsiPayload msi = GetCachedMsiPayload(msiPackageId, msiPackageVersion, null);
+            VerifyPackage(msi);
+            DetectState state = DetectPackage(msi.ProductCode, out Version installedVersion);
+
+            if (installAction == InstallAction.Install)
```

Comment:
This can be a removal as well right? I'd suggest renaming the parameter to `requestedAction`. PlanPackage can alter the requested action (or not). So after planning, if the planned action is still `InstallAction.Install` you'd be able to do this check. The code will work this way, but reading through it, it seems to be doing steps out of order